Fix NTFS SD card mount failure by adding sync and udevadm settle (BugFix)

Jefferyyen opened the pull request:
Fix an issue where Checkbox aborts during the SDHC storage test due to a race condition when mounting an NTFS-formatted SD card. The test fails with:

    ntfs-3g-mount: mount failed: Device or resource busy

because `mount_usb_storage()` executes before Ubuntu's auto-mount process completes.

Add `sync` to flush pending writes and `udevadm settle` to wait for udev events, ensuring the device is ready before mounting.

<!--
Example Title: Fixed bugged behaviour of checkbox load config (Bugfix)

A Traceability Marker is required as a suffix in the PR title to help understand the impact of your change at a glance.

Pick one of the following:
- Infra: Your change only includes documentation, comments, github actions or metabox
- BugFix: Your change fixes a bug
- New: Your change is a new backward compatible feature, a new test/test plan/test inclusion
- Breaking: Your change breaks backward compatibility.
    - This includes any API change to checkbox-ng/checkbox-support
    - Changes to PXU grammar/field requirements
    - Breaking changes to dependencies in snaps (fwts upgrade for example)

If your change is to providers it can only be (Infra, BugFix or New).

If your change impacts the submission format in Checkbox test reports, ensure that `submission-schema/schema.json` is updated and relevant fields are documented.

Signed commits are required.
  - See CONTRIBUTING.md (https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md#signed-commits-required) for further instructions.
  - If you are posting your first pull request from a fork of the repository, a Checkbox maintainer (someone with contributor / maintainer / admin rights) will be required to enable CI checks in the repo to be executed.
    - This will be communicated with a comment to the PR of the form `/canonical/self-hosted-runners/run-workflows <SHA-for-HEAD-commit>`
-->

## Description
This PR enhances the storage mounting function in Checkbox by adding sync and udevadm settle. These additions ensure that the system has fully registered the device before attempting to mount it, preventing premature failures.
<!--
Describe your changes here:

- What's the problem solved (briefly, since the issue is where this is elaborated in more detail).
- Introduce your implementation approach in a way that helps reviewing it well.
- Alert the reviewer of any changes that involve data persistence: present examples of file format changes as part of the PR description (e.g. new fields of data stored in unit or submission output).
-->

## Resolved issues
When running the **insertion + read/write + removal test**, if the SD card is formatted as **NTFS**, Checkbox may attempt to mount the device before Ubuntu's auto-mount process is complete. This results in an error log:

    ntfs-3g-mount: mount failed: Device or resource busy

which causes the test to abort prematurely.

By adding `sync` and `udevadm settle`, we ensure that the system completes the device registration before proceeding with the mounting process, preventing this issue.
<!--
Note the Jira and GitHub issue(s) resolved by this PR (`Fixes|Resolves ...`).
Make sure that the linked issue titles & descriptions are also up to date.
-->

## Documentation

<!--
Please make sure that...
- Documentation impacted by the changes is up to date (becomes so, remains so).
  - Documentation in the repository, including contribution guidelines.
  - Process documentation outside the repository.
- Tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
- When breaking changes and other key changes are introduced, the PR having been merged should be broadcast (in demo sessions, IM, Discourse) with relevant references to documentation. This is an opportunity to gather feedback and confirm that the changes and how they are documented are understood.
-->

## Tests

The fix was verified by running the following test procedure:

Test Steps:
1. Run the command:`checkbox-cli run com.canonical.certification::mediacard/sdhc-storage-manual`

2. When prompted, insert an **NTFS-formatted SD card**.
3. Observe the output logs:
```
--------- Testing insertion ---------  
INSERT NOW  
Timeout: 30 seconds  
```

4. Checkbox detects the card and attempts to mount it. Before applying this fix, the following error occurs:
```
------- Insertion test passed -------  
--------- Testing read/write --------  
DEBUG: generating a random file  
Mounting the USB storage  
sda1  
DEBUG: try to mount USB storage for testing  
ntfs-3g-mount: mount failed: Device or resource busy    🔴this one
ERROR: mount /dev/sda1 on /tmp/tmpXXXXXX failed.  
```

5. With the applied fix, `sync` and `udevadm settle` ensure that the mounting process does not fail due to a premature device state.
<!--
- How was this PR tested? Please provide steps to follow so that the reviewer(s) can test on their end.
- Please provide a list of what tests were run and on what platform/configuration.
- Remember to check the test coverage of your PR as described in CONTRIBUTING.md
-->

fernando79513 commented on .github/workflows/deb_validator.yaml:
```diff
@@ -59,8 +59,9 @@ jobs:
         run: |
-          sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 753F8DAABFD945830D46A6BB968504F7952C9377
-          sudo add-apt-repository ppa:checkbox-dev/edge
+          sudo mkdir -p /etc/apt/keyrings
+          curl -fsSL "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x753F8DAABFD945830D46A6BB968504F7952C9377" | sudo gpg --dearmor -o /etc/apt/keyrings/checkbox-dev.gpg
+          echo "deb [signed-by=/etc/apt/keyrings/checkbox-dev.gpg] https://ppa.launchpadcontent.net/checkbox-dev/edge/ubuntu $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/checkbox-dev.list
```
Why do you need this change? 
I think the issue is with your GPG keys not being properly verified.

Jefferyyen replied:
I encountered a timeout error with `sudo add-apt-repository ppa:checkbox-dev/edge` in CI/CD, so I used an alternative method to fetch the key to avoid blocking my commit.
Here is the log, did i get it wrong and need to reset/rebase this commit?

```
deb_validation (ubuntu-20.04, providers/sru)
failed 1 hour ago in 2m 16s

Run sudo add-apt-repository ppa:checkbox-dev/edge
 
To align our distribution channels, have transitioned to a new set of PPAs.
These new channels will replace our existing ones and provide a clearer, more
consistent structure for our releases.

------------------------------------------------------------------------------
Edge: ppa:checkbox-dev/edge: For testing recent developments not yet in beta.
------------------------------------------------------------------------------

To install Checkbox testing suite, please do the following:

sudo add-apt-repository ppa:checkbox-dev/edge
sudo apt-get update
sudo apt-get install canonical-certification-client

You can then run Checkbox by running:
- checkbox-cli (textual user interface)
 More info: https://launchpad.net/~checkbox-dev/+archive/ubuntu/edge
Error: retrieving gpg key timed out.
Error: Process completed with exit code 1.

```

fernando79513 commented on checkbox-support/checkbox_support/scripts/tests/test_mount_usb_storage.py:
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
```
I think this file is in the propper place, but I don't see a reason why it is separated from the `test_usb_read_write.py` script.
I would add the contents of that test to this one, and rename this one to `test_usb_read_write.py`

Jefferyyen replied:
I've committed the changes as suggested. I'm not sure if my understanding is completely correct, please let me know if anything needs further adjustment. Thank you.

fernando79513 replied:
I think I didn't make it super clear with my comment. The first part was fine (rename the test to `test_usb_read_write.py` under `checkbox_support/scripts/tests`
There was also another `test_usb_read_write.py` under  `checkbox_support/tests` that you have now merged. You could just remove this one, since it's already merged with the one in `checkbox_support/scripts/tests/test_usb_read_write.py`
This was an issue in our part because the test was not in the propper place to begin with, but thank you for the fix!

fernando79513 reviewed (commented):
Thank you very much for this PR!
It's really well documented and explained. I've just added a couple of comments.
The main issue is that I think you have to validate your gpg keys. That will fix the issue with the deb validator and with the merge of the PR.

Jefferyyen commented on checkbox-support/checkbox_support/scripts/tests/test_mount_usb_storage.py:
```diff
@@ -0,0 +1,106 @@
+                return 0  # Other commands also return success
+
+        mock_call.side_effect = call_side_effect
+
```
Thanks for pointing that out! I've updated the code as suggested.

fernando79513 commented on checkbox-support/checkbox_support/scripts/tests/test_usb_read_write.py:
```diff
@@ -0,0 +1,210 @@
+        # Ensure that sys.exit was not called
+        mock_exit.assert_not_called()
+
+    @patch("checkbox_support.scripts.usb_read_write.subprocess.call")
+    @patch(
+        "checkbox_support.scripts.usb_read_write.os.path.join",
+        return_value="/dev/sda1",
+    )
+    @patch(
+        "checkbox_support.scripts.usb_read_write.sys.exit",
+        side_effect=SystemExit,
+    )
+    def test_mount_usb_storage_failure(self, mock_exit, mock_join, mock_call):
+        """
+        Test the failure scenario:
+          - Simulate that the mount command returns 1 (failure).
+          - The program should call sys.exit(1) and raise a SystemExit.
+        """
+
+        def call_side_effect(args, **kwargs):
+            if args[0] == "mount":
+                return 1  # Simulate mount failure
+            else:
+                return 0
+
+        mock_call.side_effect = call_side_effect
+
+        with patch(
+            "checkbox_support.scripts.usb_read_write.FOLDER_TO_MOUNT",
+            "/mnt/usb",
+        ):
+            with self.assertRaises(SystemExit):
+                with mount_usb_storage("sda1"):
+                    # As soon as we enter the context, a non-zero return from mount should trigger sys.exit(1)
+                    pass
+
+        # Check that sys.exit was called correctly (with 1)
+        mock_exit.assert_called_once_with(1)
```
I think this test is more complex than it has to be. You don't have to mock the `sys.exit()` or the `join()` functions, it should be just okay mocking the subprocess. You could also check that the right message has been logged though.
```suggestion
    @patch("checkbox_support.scripts.usb_read_write.subprocess.call")
    @patch("logging.error")
    def test_mount_usb_storage_failure(self, mock_log, mock_call):
        """
        Test the failure scenario:
          - Simulate that the mount command returns 1 (failure).
          - The program should call sys.exit(1) and raise a SystemExit.
        """

        def call_side_effect(args, **kwargs):
            if args[0] == "mount":
                return 1  # Simulate mount failure
            else:
                return 0

        mock_call.side_effect = call_side_effect

        with patch(
            "checkbox_support.scripts.usb_read_write.FOLDER_TO_MOUNT",
            "/mnt/usb",
        ):
            with self.assertRaises(SystemExit) as context:
                with mount_usb_storage("sda1"):
                    # As soon as we enter the context, a non-zero return from
                    # mount should trigger sys.exit(1)
                    pass

        self.assertEqual(context.exception.code, 1)

        # Verify that the program logs the error message
        mock_log.assert_called_once_with(
            "mount /dev/sda1 on /mnt/usb failed."
        )
```

Jefferyyen commented:
Thank you for your suggestion. I've committed the changes to the PR again. Please don't hesitate to let me know if anything needs to change.

fernando79513 commented:
As mentioned [here](https://github.com/canonical/checkbox/pull/1717#discussion_r1971190946), please try to remove the changes to the GitHub actions, these shouldn't be needed after the latest fixes, If they still don't work, try rebasing your branch with main again.

Jefferyyen commented:
Thanks for the feedback. I've removed the GitHub Actions changes and deleted the redundant test file. Please let me know if you have any further suggestions.
